### ISSUE TITLE
job stuk waiting for PJ

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5876]
I have digged a bit into one situation in which one job in a task completed since one day, but is still waiting for PostJob to run. Which indeed happens every now and then.
I found that DAGMAN did not notice that the job completed, so it is still waiting and not running the postjob.
I had seen the same already a few times, sounds like DAGMAN bug.
I sent a bug report via mail to Jaime (below).
This issue is for my reminder and in case the solution requires us to do something

```
-------- Forwarded Message --------
Subject: possible bug in DAGMAN
Date: Thu, 18 Apr 2019 12:31:55 -0500
From: Stefano Belforte <stefano.belforte@cern.ch>
To: Jaime Frey <jfrey@cs.wisc.edu>
CC: stefano.belforte@cern.ch

Hi Jaime
sorry for direct mail, I do not seem to have better ways to report this.
I have found a couple occasions where it looked like DAGMAN was loosing
track of some node, there may be more of course, so this time I thought
to report it.
Since it is very rare, it is not a real problem for us, but in case
you want to find out what's happening... you may want to open a tkt

I have this DAG with 5508 nodes, each of them is a simple
prejob/vanillajob/postjob (the CMS CRAB standard)
For node 5044 I find in DAGMAN status file (which we call node_state)
[
  Type = "NodeStatus";
  Node = "Job5044";
  NodeStatus = 3; /* "STATUS_SUBMITTED" */
  StatusDetails = "not_idle";
  RetryCount = 0;
  JobProcsQueued = 1;
  JobProcsHeld = 0;
]

and in dagman log
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$ grep 28720855 RunJobs.dag.nodes.log
000 (28720855.000.000) 04/17 02:25:58 Job submitted from host: <188.184.85.27:4080?addrs=188.184.85.27-4080&noUDP&sock=1426426_7bc8_3>
001 (28720855.000.000) 04/17 02:45:57 Job executing on host: <131.225.191.224:33998?CCBID=137.138.148.95:9618%3faddrs%3d137.138.148.95-9618%26noUDP%26sock%3dcollector9671#600617%20131.225.205.232:9639%3faddrs%3d131.225.205.232-9639#904863&addrs=131.225.191.224-33998+[--1]-33998&noUDP>
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$

and in DAGMAN stdout:
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$ grep 28720855 RunJobs.dag.dagman.out |tail -2
04/18/19 19:01:06   Node Job5044, HTCondor ID 28720855, status STATUS_SUBMITTED
04/18/19 19:11:09   Node Job5044, HTCondor ID 28720855, status STATUS_SUBMITTED
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$


But that job completed one day ago:
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$ condor_history -userlog job_log 28720855
 ID     OWNER          SUBMITTED   RUN_TIME     ST COMPLETED   CMD
28720855.0   cms1627         4/17 02:25   0+00:07:05 C   4/17 02:53 ???

same if I query global logs instead of the userlog
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$ condor_history --match 1 28720855
 ID     OWNER          SUBMITTED   RUN_TIME     ST COMPLETED   CMD
28720855.0   cms1627         4/17 02:25   0+00:07:16 C   4/17 02:53 /data/srv/glidecondor/condor_lo
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$ condor_q 28720855


-- Schedd: crab3@vocms0106.cern.ch : <188.184.85.27:4080?... @ 04/18/19 19:31:33
 ID      OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD

0 jobs; 0 completed, 0 removed, 0 idle, 0 running, 0 held, 0 suspended
[cms1627@vocms0106 cluster28708598.proc0.subproc0]$


Why does DAGMAN still think that it is running ?

I believe you have access to vocms0106 for investigating, the directory is
/data/srv/glidecondor/condor_local/spool/8598/0/cluster28708598.proc0.subproc0

and I am certainly and always available to help if useful.

Thanks
Stefano
```
I also added this info:
```



-------- Forwarded Message --------
Subject: Re: possible bug in DAGMAN
Date: Thu, 18 Apr 2019 12:49:06 -0500
From: Stefano Belforte <stefano.belforte@cern.ch>
To: Jaime Frey <jfrey@cs.wisc.edu>
CC: stefano.belforte@cern.ch

clarification:
we need the PostJob script to run for all CRAB jobs, but
here DAGMAN is not running it. Which is how I noticed it and
why it is a problem).

Current workaround is for user to kill and resubmit in CRAB
which forces the DAGMAN to rerun this node via its rescue mode.
But I do not need this DAG to be 100% complete now, so am
leaving it as is in case there's some useful info in logs.
That directory should stay around for 4 more weeks.

Stefano

```

**Comment by belforte** [comment]
HTCondor side problem acknowledged: https://htcondor-wiki.cs.wisc.edu/index.cgi/tktview?tn=7008

**Comment by belforte** [comment]
Update from Jamie:
```
Thanks Jamie,
I agree that it seems like a failure of the underlaying OS,
nothing that you can prevent, only a matter of whether it is worth
to try to detect it after it happened

I will let you know if this happens again afger we have
upgraded to 8.8.4 (IIUC the ticket)

Stefano

On 11/06/2019 10:45, Jaime Frey wrote:
>> On Apr 18, 2019, at 2:47 PM, Stefano Belforte <stefano.belforte@cern.ch> wrote:
>>
>> thanks. If it comes that you need input from me, up to you
>> whether to use mail of give me write access.
>>
>> On 18/04/2019 14:15, Jaime Frey wrote:
>>> I’ve made a ticket in our system:
>>> https://htcondor-wiki.cs.wisc.edu/index.cgi/tktview?tn=7008
>
>
> I’ve pushed a change in the code to detect when it tries to write a job event and the OS fails to write the full text. I think we’ll have to leave at that unless we see more occurrences. I don’t see any bugs in the Condor code that would trigger the behavior you saw.
>
>   - Jaime
>

```

**Comment by belforte** [comment]
one year later...

>  -------- Forwarded Message --------
Subject: job terminate event missing in Dagman log
Date: Fri, 18 Sep 2020 16:02:01 +0200
From: Stefano Belforte <stefano.belforte@cern.ch>
To: Marco Mascheroni <marco.mascheroni@cern.ch>, Jaime Frey <jfrey@cs.wisc.edu>
CC: stefano.belforte@cern.ch, Antonio Perez-Calero Yzquierdo <aperez@pic.es>, James R Letts <letts@physics.ucsd.edu>, Edgar Fajardo Hernandez <emfajardohernandez@physics.ucsd.edu>, Todd Tannenbaum <tannenba@cs.wisc.edu>, Jeff Dost <jdost@ucsd.edu>, Farrukh Aftab Khan <farrukh.aftab.khan@cern.ch>, Edgar Fajardo Hernandez <emfajard@ucsd.edu>, Dirk Hufnagel <Dirk.Hufnagel@cern.ch>, Ben Jones <ben.dylan.jones@cern.ch>, Todor Trendafilov Ivanov <todor.trendafilov.ivanov@cern.ch>, Alan Malta Rodrigues <alan.malta@cern.ch>, Saqib Haleem <saqib.haleem@ncp.edu.pk>, Marco Mambelli <marcom@fnal.gov>, Sharad Agarwal <sharad.agarwal@cern.ch>, prajesh sharma <prajesh.sharma@cern.ch>, John M Knoeller <johnkn@cs.wisc.edu>, Diego Davila <diego.davila@cern.ch>, Edita Kizinevic <Edita.kizinevic@cern.ch>, Daniele Spiga <daniele.spiga@cern.ch>, Maria P. Acosta <macosta@fnal.gov>, Krista Majewski <klarson1@fnal.gov>, Dave Mason <dmason@fnal.gov>, Brian Bockelman <brian.bockelman@cern.ch>, Kenyi Paolo Hurtado Anampa <khurtado@nd.edu>

Dear developers,
this keeps happening
https://htcondor-wiki.cs.wisc.edu/index.cgi/tktview?tn=7008
e.g. [*]

I know that Jaime tried hard and log to add diagnostics to catch it in
the act, w/o success. But while reasonably rare, it is annoying, right ?
Could you consider re-opening the ticket (or create a new one) and
possibly evaluate the possibility that DAGMAN sort of re-sync itself
with job log sort of once or twice a day ?

Just an idea...
Stefano



[*]
belforte@vocms0122/SPOOL_DIR> pwd -P
/data/srv/glidecondor/condor_local/spool/7744/0/cluster60977744.proc0.subproc0
belforte@vocms0122/SPOOL_DIR> grep 61014901 job_log |tail -4
Cluster = 61014901
005 (61014901.000.000) 09/08 10:21:22 Job terminated.
028 (61014901.000.000) 09/08 10:21:22 Job ad information event triggered.
Cluster = 61014901
belforte@vocms0122/SPOOL_DIR> grep 61014901 RunJobs.dag.nodes.log
000 (61014901.000.000) 09/07 08:27:26 Job submitted from host: <188.184.95.191:4080?addrs=188.184.95.191-4080&noUDP&sock=48858_1a46_272>
001 (61014901.000.000) 09/07 19:43:49 Job executing on host: <131.154.208.182:40094?CCBID=137.138.148.95:9618%3faddrs%3d137.138.148.95-9618%26alias%3dvocms0806.cern.ch%26noUDP%26sock%3dcollector9686#2534862%20131.225.188.15:9694%3faddrs%3d131.225.188.15-9694%26alias%3dcmssrv601.fnal.gov#1912794&addrs=131.154.208.182-40094&noUDP>
belforte@vocms0122/SPOOL_DIR>

**Comment by belforte** [comment]
there's been progress in understanding this over last days. Most likely the problem is in CRAB and solutions have been indicatted.
Ttime to get serious about it. So I changed prioritiy.
Issue still On Hold awaiting the chance to talk with Marco M. about best strategy.

**Comment by belforte** [comment]
I collect here recent mails on this from the cm-htcondor-developers mailing list

```


On 24/09/2020 00:14, Jaime Frey wrote:
> I think we’ve found a smoking gun. Most of the missing terminate events I’ve found so fare happen when the condor_dagman process is restarted. Do the wrapper scripts around condor_dagman do any editing of the job even log files?
> 
>   - Jaime
> 
>> On Sep 23, 2020, at 3:49 PM, Jaime Frey <jfrey@cs.wisc.edu> wrote:
>>
>> I’ve written a script to find all instances of missing terminate events for currently-queued jobs. It turned up a couple more-recent occurrences on vocms0122. Recent enough to still have daemon logs, which don’t show anything informative. Mark, I’ll gather everything related to one of them for you to peruse.
>>
>> - Jaime
>>
```
```

On 24/09/2020 09:36, Marco Mascheroni wrote:
> Hi Jaime,
> 
> we actually do some editing of the dagman.nodes file:
> 
> https://github.com/dmwm/CRABServer/blob/master/scripts/AdjustSites.py#L33
> 
> According to the funcion comments:
> 
> In practice, search for this kind of
> 
> sequence in the DAG .nodes.log file:
> 
> ...
> 
> 016 (146493.000.000) 11/11 17:45:46 POST Script terminated.
> 
> (1) Normal termination (return value [0|2])
> 
> DAG Node: Job105
> 
> ...
> 
> for the job ids in resubmitJobIds and replace the return value to 1.
> 
> If resubmitJobIds = True, only replace return values 2 (not 0) to 1.
> 
> The edited .nodes.log file is written into a .nodes.log.tmp file and then moved to the original one. I guess that if condor modifies the file at the same time some updates might go missing... We usually lock files with fcntl when there are race conditions: https://github.com/dmwm/CRABServer/blob/master/src/python/ServerUtilities.py#L379-L385
> 
> But I am not sure it would help in this case..
> 
```
```


On 24/09/2020 15:57, Jaime Frey wrote:
> At six years old, this code was written when HTCondor locked the event logs for all writing. Re-enabling the locking should be a safe change to fix this problem.
> 
>   - Jaime
> 
>> On Sep 24, 2020, at 8:52 AM, Stefano Belforte<stefano.belforte@cern.ch>  wrote:
>>
>> I will be gone next week. Marco can you think about this a bit and once
>> I am back we turn this into something that (even) I can realistically do
>> and capture it in a GH issue ? I am very hesitant to say the least to
>> put my hands in code not touched since 6 years and which
>> I never read which executes an algorithm which I never really understood.
>>
>> I prefer not to comment now on lock/unlock/append as I do not understand enough
>> to say intelligent things.
>>
>> Thanks all !! Apparently we have made an enormous progress here !
>>
>> Stefano
>>
>>
>> On 24/09/2020 15:35, Jaime Frey wrote:
>>> This may have been a safe thing to do when first deployed. Before HTCondor 8.3.6, the condor daemons locked the event logs when writing new events. Starting in 8.3.6, we disable the locking by default, open the files in append mode, and write each event as a single write() call. You can enable locking of the event logs by setting ENABLE_USERLOG_LOCKING=true in the config file.
>>> You should also consider changing your code to append to the .nodes.log file directly (in append mode) instead of using a temporary file (and not do locking).
>>>   - Jaime
>>>> On Sep 24, 2020, at 3:14 AM, Stefano Belforte <stefano.belforte@cern.ch  <mailto:stefano.belforte@cern.ch>> wrote:
>>>>
>>>> thanks Marco. IIUC that code is only executed when DAGMAN is restarted
>>>> after a user issues crab resubmit. So timing would match.
>>>> I suspect that when this resubmit logic was designed you and others
>>>> where not aware that condor daemons would be writing to this file even
>>>> if dagman itself is not running at that time.
>>>>
>>>> Stefano
```
followed by an errata corrige:
```
Sorry, this is was what my final sentence should be:

You should also consider changing your code to overwrite the job exit code in the terminate events in-place in the live .nodes.log file instead of using a temporary file (and not do locking).

 - Jaime

```

**Comment by belforte** [comment]
and finally this looks like final suggestion from Jaime for us to try:
```
One more update, now that I’ve actually read the relevant CMS code. I see now that it reads the whole .nodes.log file into memory, editing some postscript exit codes as it goes. It then re-opens the .nodes.log file in truncate/write mode and writes everything back out. This will definitely lead to lost events. Using mode “r+” instead of “w” for python’s os.open() call should remove the need for locking. It does require that the edits don’t change the length of the file.

 - Jaime
```

**Comment by belforte** [comment]
```
On 25/09/2020 05:14, Jaime Frey wrote:
>> On Sep 24, 2020, at 4:29 PM, Stefano Belforte <stefano.belforte@cern.ch> wrote:
>>
>> One request, can you share your script which looks for such
>> incidents ? So we can use it to test code changes.
> 
> 
> You can find the script here: /afs/cern.ch/user/j/jafrey/public/searcher.sh
```
and here's the script
```bash
#!/bin/sh

# Search the spool directory of a CRAB schedd for instances where a
# terminate event is in the user job log but not the dagman event log.

spool=/data/srv/glidecondor/condor_local/spool
job_tmp=job_tmp.$$
dag_tmp=dag_tmp.$$
diff_tmp=diff_tmp.$$

for dir in ${spool}/[0-9]*/0/cluster*0 ; do
  #echo $dir
  grep '^00[015]' $dir/job_log | cut -d' ' -f1,2 | sort >$job_tmp
  cat $dir/RunJobs*.nodes.log | grep '^00[015]' | grep -v 2147483647 | cut -d' ' -f1,2 | sort >$dag_tmp
  if ! diff $job_tmp $dag_tmp >$diff_tmp ; then
    echo $dir
    cat $diff_tmp
  fi
done

#rm -f $job_tmp $dag_tmp $diff_tmp

```

**Comment by belforte** [comment]
one more note:
```
Mode “w” truncates the file on open. Mode “r+” opens the file for writing without truncating it. If you use mode “r+”, you overwrite the existing file contents with identical data except for the modified exit codes, and the Condor daemons are still free to append new events at the end of the file.

 - Jaime

> On Sep 25, 2020, at 4:07 AM, Marco Mascheroni <marco.mascheroni@cern.ch> wrote:
>
> Hi Jaime,
>
> I do believe the len of the file does not change (assuming we are putting the right amount of spaces here: https://github.com/dmwm/CRABServer/blob/master/scripts/AdjustSites.py#L90)
>
> anyway, can you elaborate on this:
>
>     Using mode “r+” instead of “w” for python’s os.open() call should remove the need for locking. It does require that the edits don’t change the length of the file.
>
>
> I don't think I am getting all the details and why using 'r+' instead of 'w' will help (Well, worst case we will discuss this next Wed at the condor devs+CMS meeting)
>
>
```

**Comment by belforte** [comment]
discussed a bit with Marco in private mail thread.
Looks very promising but we need to test well to be sure of what we do. in particular the number of bytes read and written match. And surely test on a real schedd etc.
May be good to hand to someone who wants to do some python practice and can invest the needed time... student ? operator ?

**Comment by belforte** [comment]
After discussion with HTCondor developers we came to these conclusion:

1. the problem was not there in 2014 when Brian initially coded this
1. the problem came when condor got smarter about writing logs and stopped locking them by default, opening the way for our log editing procedure (which attempts to use condor file locking API) to overwrite a log file with an "old" version where some events are missing
1. condor can be configured to revert to old behavior by setting `ENABLE_USERLOG_LOCKING=True` in its configuration so that we do not have the problem
1. we made that change in all CRAB schedd's and did not find any sign of increased load or slower operations, so we can run in that way for a while. Ref. https://cms-logbook.cern.ch/elog/Analysis+Operations/3282 
1. editing logs is bad anyhow and it has been agreed to enhance DAGMAN functionality to allow CRAB to do resubmits w/o tampering with files it should not tamper with. Discussion on this has started with HTCondor DAGMAN expert Mark Coatsworth:
 https://docs.google.com/document/d/1vgJApmjkH9brYhQZbdRooGnj2mqSrme7BWVxzReZ0oM/edit
